### PR TITLE
ENH: Interpolate blinks in eyetrack channels

### LINKED ISSUE #11725: Feature Request: Interpolate blinks in eyegaze / pupil channels.
### Describe the new feature or enhancement

While I wrap up my open PR soon, I wanted to initiate a discussion on the next task that I'd like to work on for my GSoC project.

CC @larsoner and @britta-wstnr 

#### Proposed function: 

`mne.preprocessing.interpolate_blinks` (or `remove_blink_artifact`). For `pupil` channels

#### Background 
 it is a common pre-processing step in pupillometry research to interpolate missing values during blinks (see [this paper](https://www.researchgate.net/profile/Sebastiaan-Mathot/publication/236268543_A_simple_way_to_reconstruct_pupil_size_during_eye_blinks/links/00b7d51779b2e5e567000000/A-simple-way-to-reconstruct-pupil-size-during-eye-blinks.pdf), [this paper](https://journals.sagepub.com/doi/10.1177/2331216518800869), [this package](https://cran.r-project.org/web/packages/PupilPre/vignettes/PupilPre_Interpolation_and_Filtering.html), and [this package](https://github.com/pyeparse/pyeparse)).

The basic idea is that blinks are very short in duration, but the pupil response is relatively slow; so it can be reasonably assumed that the pupil size during a blink was close to the pupil size values just before and just after the blink.

#### Use-case
Currently in `mne`, if you have a `RawEyelink` object with `pupil` channels, and you want to to assess the average pupil size across  trials, you have to:

- epoch the data
-  drop any `epochs` that have a blink in them (which can result in data loss). 
- average aross the remaining `epochs`

This is because  `pupil` channels contain `0`'s during blinks. Including blink periods in `epochs.average()` would bias the result such that the the pupil size would seem to be unrealistically small!

Having an `interpolate_blinks` function would help users move beyond a `Raw` object to epoching the data and being able to use those epochs for analyses.

### Describe your proposed implementation

I propose we add a function, called `interpolate_blinks` or `remove_blink_artifact` to `mne.preprocessing.eyetracking`.

#### How:
The way I am thinking about it now:

- This function would accept a `Raw` object, and will rely on `Annotations` in the `Raw` object that start with `blink` to identify periods to be interpolated.
   - For example, `mne.io.read_raw_eyelink` extracts any blink periods (that EyeLink wrote to the file) and stores them as `mne.Annotations`. So this function would rely on those `Annotations`.
- This function would accept a parameter named something like `method`, which would accept `'linear'` or `'cubic_spline'`, and would use:

    - `numpy.interp` for linear interpolation
    - `scipy.interpolate.CubicSpline` for cubic-spline interpolation (see [this eye-tracking paper](https://www.researchgate.net/profile/Sebastiaan-Mathot/publication/236268543_A_simple_way_to_reconstruct_pupil_size_during_eye_blinks/links/00b7d51779b2e5e567000000/A-simple-way-to-reconstruct-pupil-size-during-eye-blinks.pdf)).

#### Summary:
So I suggest that: 

- This function will only operate on  `pupil` channels (not `eyegaze`, `EEG`, `MEG`, `NIRS` etc.).
- This function can start with interpolating data during _blinks_. Then maybe we can extend this to interpolating `pupil` data during _saccades_  (There is precedent for this because pupil-size measurement is thought to be less reliable during saccades).
- Then, _maybe_ we can extend this function to `eyegaze` channels  (There is [one package that does this](https://cran.r-project.org/web/packages/PupilPre/vignettes/PupilPre_Interpolation_and_Filtering.html)), but AFAIK it is harder to justify interpolating `eyegaze` data, because it can change at a much faster rate / is more variable across time. I'll need to do more literature review to see how often this has been done.

### Describe possible alternatives

I'm open to any other ideas!

### Additional context

To see an example of a file that could benefit from `interpolate_blinks` see:

```
from mne.datasets.testing import data_path
from mne.io import read_raw_eyelink
fname = data_path() / "eyetrack" / "test_eyelink.asc"
raw = read_raw_eyelink(fname, find_overlaps=True, create_annotations=['blinks'])
raw.plot(scalings='auto')
```
test_eyelink_plot

### PR DESCRIPTION
**Pull request by scott-huberty**
#### Reference issue
Closes #11725 

#### What does this implement/fix?
This implements a linear interpolation function in `mne.preprocessing.eyetracking`, to interpolate missing values in `pupil` channels during blinks. 

This time I wanted to get this PR up early / in a minimal working-state, so that devs have a change to review the foundation of the code, before we add more to it next week (see additional info below). I'm definitely interested in your feedback on the code design, if there are better (or more _mne-like_ ) approaches you think I could take, I'm happy to hear your ideas🙂 

#### Additional information

After discussing this function yesterday with @britta-wstnr , @mscheltienne , and @drammock , here are some ideas that were suggested (feel free to let me know if i missed anything):

- [x] remove `pytest.mark.xfail` from the test for this module, after #11746 is merged and confirm that tests pass
- [x] make the interpolation channel specific. Right now, in the case of a binocular recording (2 `pupil` channels), if there were a "wink" in the recording, **both** `pupil` channels would be interpolated.

     - [x] to do this, we should refactor `read_raw_eyelink` to add channel information in the `blink` annotations.
     - [x]  There was also a discussion of whether `read_raw_eyelink` should prepend blink annotations with `BAD_`, to be consistent with other mne functions. Then, this function would _only_ interpolate over `'BAD_blink'` annotations by default. I'm not sure if there was a resolution on this matter, and I'm not sure yet how I feel about the idea.
 
- [x] allow the user to specify the `annotation` descriptions that should be considered as blinks? something like `blink_annotations=['my_manual_blink'`] ? we can give it a default value of `None`, which will use _any_ `annotation` description that starts with `'blink'`.

also CC @larsoner 

#### Minimally Working example

```
import mne
fname = mne.datasets.testing.data_path() / "eyetrack" / "test_eyelink.asc"

raw = mne.io.read_raw_eyelink(fname, create_annotations=["blinks"],
                              find_overlaps=True, preload=True)
mne.preprocessing.eyetracking.interpolate_blinks(raw, buffer=0.05)
raw.pick(["pupil_left", "pupil_right"])
raw.plot()
```
interpolate

Here is the updated [eye-tracking tutorial](https://output.circle-artifacts.com/output/job/85641281-1930-473c-8c02-d262dbe2bca9/artifacts/0/html/auto_tutorials/preprocessing/90_eyetracking_data.html)

### CODE REVIEW

**Comment by drammock** [comment]
> * to do this, we should refactor `read_raw_eyelink` to add channel information in the `blink` annotations

This is not a necessary step in this PR (though we should do it). You can make this function channelwise-annotation-aware without changing the reader's behavior.

> * whether `read_raw_eyelink` should prepend blink annotations with `BAD_`, to be consistent with other mne functions

IIRC the place we left this was "don't make the reader too smart/powerful, read in what's there in the file and then the user can use `annotations.rename` if they want

> allow the user to specify the `annotation` descriptions that should be considered as blinks

I think this is a necessity. I'm not sure if list of strings, regex, callable, etc is best here

**Comment by mscheltienne** [comment]
> > This is not a necessary step in this PR (though we should do it). You can make this function channelwise-annotation-aware without changing the reader's behavior.

> The other way that I can think of doing this is to rely on the eye name being in the annotation description. But if we think that relying on the 'ch_names' key of the annotation is the best way, I don't mind doing a separate PR to update read_raw_eyelink.

Relying on the `ch_names` key is cleaner. The field exist to create channel-wise annotations. IMO, storing this information as part of the description will be harmful in the long run.

-------

For `blink` vs `BAD_blink`, I agree that the reader should not be too smart/powerful. The "smarter" you make it, the more likely it will break on an edge case. Reading information without transforming it is always simpler. Moreover, `BAD_blink` would make sense since data isn't usable during a blink, but it would create a discrepancy with the other annotations/events read from the same file. 

I would however mention clearly in the tutorial that you can rename the loaded annotations, with an example code snippet.  

--------

> eyegaze is np.nan and pupil is 0

This is now sparking another question, how does MNE handle a segment during which some of the channels are `np.nan`?
It's not a big issue for a raw containing only eyelink channels, but once merged/synchronized with a MEG recording, it opens access to many functions. Will this `np.nan` segment on the eyegaze channel cause issues? Should it be clearly annotated with a channel-wise annotation?

**Comment by scott-huberty** [comment]
>  Will this `np.nan` segment on the eyegaze channel cause issues? Should it be clearly annotated with a channel-wise annotation?

There are a couple places where there could be issues (that I've seen):
- Epochs: if you include blink periods in your `epochs`, if you call `epochs.average()`, an error will be thrown, because `average` doesn't use `np.nanmean`.
- Filtering.  I don't think the `filter` function ignores `eyegaze` channels so you have to make sure you exclude them via the `picks` param. Otherwise IIRC the `np.nan`'s will cause the filter to do strange things to those signals. I assumed that `filter` function would only act on "data channels"

**Comment by larsoner** [comment]
@scott-huberty I have not read all discussion but is a path forward clear? If not, want me to read through and comment and/or we chat at some point?

**Comment by scott-huberty** [comment]
> @scott-huberty I have not read all discussion but is a path forward clear? If not, want me to read through and comment and/or we chat at some point?

I think so:

- Make this function channelwise-annotation aware
- In a separate PR, update `read_raw_eyelink` to store channelwise info for occular annotations.. So that this function can be easily used with data loaded from `read_raw_eyelink`.
- We require the user to pass in the description(s) of blink anntoations?

**Comment by drammock** [comment]
> We require the user to pass in the description(s) of blink anntoations?

To me the API of this part is still not clear. Possible inputs to a `annotations_to_interpolate` param might be:

- string
- sequence of strings
- callable that returns `True`/`False`
- dict that maps annotation descriptions to `True`/`False`
- regex

I'd like @larsoner's perspective on this, I feel like his YAGNI radar is better than mine. (also we need a better param name than `annotations_to_interpolate`, maybe `match`?)

**Comment by larsoner** [comment]
I would start with `str | list of str` because usually it's just going to be `BAD_blink` (or whatever we read it as by default) and adding `list of str` is easy once you have `str`. And if people really have 5+ kinds they want to interpolate we can add `re` support but I doubt we'll get there

**Comment by scott-huberty** [comment]
@larsoner and I met this afternoon and here's a summary of our plan (for anyone to confirm or object to):

- In a separate PR, update `read_raw_eyelink` to store channelwise info for occular annotations (this should be pretty easy/quick 🤞) .
- `read_raw_eyelink` should prepend `BAD_` to blink annotations, because, upon reading them, the data within those times are not usable.
- `interpolate_blinks` will be channelwise-annotation aware and will act on both `eyegaze` and `pupil` data.
- `interpolate_blinks` will have a `match` parameter that has a default value of `BAD_blink`, but users can pass in another description.
- If the annotation descriptions started with `'BAD_'` (i.e. `'BAD_blink'`), After interpolating the `'BAD_'` will be removed (i.e. `blink`), because the data are now usable. This is akin to how channel interpolation functions will remove the interpolated channels from `info['bads']`

**Comment by larsoner** [comment]
... and just to add one more point: some of this stuff (e.g., removing the `BAD_` prefix, interpolating eyegaze in addition to pupil data, etc.) could in principle be made optional later. But thinking about the 95%+ of use cases we decided it shouldn't be necessary, so let's consider making this stuff optional if people ask for it.

File: mne/preprocessing/eyetracking/eyetracking.py [hunk 1]
```diff
@@ -5,7 +5,100 @@
...
+
+    Parameters
+    ----------
+    raw : instance of Raw
+        The raw data with at least one ``'pupil'`` or ``'eyegaze'`` channel.
+    buffer : float | array-like of float, shape ``(2,))``
+        The time in seconds before and after a blink to consider invalid and
+        include in the segment to be interpolated over. Default is ``0.05`` seconds
+        (50 ms). if array-like, the first element is the time before the blink and the
+        second element is the time after the blink to consider invalid, for example,
```
Comment (britta-wstnr) [inline mne/preprocessing/eyetracking/eyetracking.py:27]:
```suggestion
        (50 ms). If array-like, the first element is the time before the blink and the
        second element is the time after the blink to consider invalid, for example,
```

File: mne/preprocessing/eyetracking/eyetracking.py [hunk 2]
```diff
@@ -5,7 +5,100 @@
...
+        ``(0.025, .1)``.
+    match : str | list of str
+        The description of annotations to interpolate over. If a list, data within
+        annotations matching any of the strings in the list will interpolated over.
```
Comment (britta-wstnr) [inline mne/preprocessing/eyetracking/eyetracking.py:31]:
```suggestion
        The description of annotations to interpolate over. If a list, the data within
        all annotations that match any of the strings in the list will be interpolated over.
```
I gave this a small update - as a non-native speaker I found the sentence hard to parse, maybe this is clearer/easier to read?

File: mne/preprocessing/eyetracking/eyetracking.py [hunk 3]
```diff
@@ -5,7 +5,100 @@
 
 import numpy as np
 
+from ...io import BaseRaw
 from ...io.constants import FIFF
+from ...utils import logger, _check_preload, _validate_type, warn
+
+
+def interpolate_blinks(raw, buffer=0.05, match="BAD_blink"):
```
Comment (britta-wstnr) [inline mne/preprocessing/eyetracking/eyetracking.py:13]: Should this also take channels? Is there any scenario where someone would like to interpolate for one channel but not the other?

File: mne/preprocessing/eyetracking/eyetracking.py [hunk 4]
```diff
@@ -5,7 +5,100 @@
...
+        mask = np.zeros_like(raw.times, dtype=bool)
+        for annot in blink_annots:
+            if "ch_names" not in annot or not annot["ch_names"]:
+                msg = "blink annotation missing 'ch_names' key. got: {}".format(annot)
+                raise ValueError(msg)
+            start = annot["onset"] - pre_buffer
+            end = annot["onset"] + annot["duration"] + post_buffer
+            ch_names = annot["ch_names"]
+            if ch_info["ch_name"] not in ch_names:
+                continue  # skip if the channel is not in the blink annotation
```
Comment (britta-wstnr) [inline mne/preprocessing/eyetracking/eyetracking.py:87]: I don't really work with annotations much, so please excuse my ignorance here ... I was wondering:
a) is there any scenario where annotations would only exist for one channel? E.g. based on how additional preproc was done?
b) if a), is there an easy way to "broadcast" annotations to other channels?
c) if not b), should we allow some way to include channels and treat them based on the annotation of other channels?

**Review by britta-wstnr — COMMENTED** [review]
Looks like this is coming along great! I fixed some typos and had some higher-level questions - see below.

File: mne/preprocessing/eyetracking/eyetracking.py (hunk 4 above)
Comment (drammock) [reply mne/preprocessing/eyetracking/eyetracking.py:87]: > is there any scenario where annotations would only exist for one channel? 

The EyeLink data files can specify "blink_L" / "blink_R" / "blink_BOTH" when recording in binocular mode. Accordingly, our reader has been updated to annotate L and R channels separately or jointly.

> is there an easy way to "broadcast" annotations to other channels?

No. `annotations.ch_names` must either be `None` (all annotations are channel-agnostic) or a (ragged) array of channel names (all annotations are channel-specific; channel-agnostic spans will be a list of all channel names).

> should we allow some way to include channels and treat them based on the annotation of other channels?

My hunch here is "YAGNI", WDYT?

Comment (britta-wstnr) [reply mne/preprocessing/eyetracking/eyetracking.py:87]: Thanks for the explanation @drammock, now I understand it better I think!
Mh ... what if you have "blink_L" but you want to interpolate the data in left _and_ right channels?

Comment (larsoner) [reply mne/preprocessing/eyetracking/eyetracking.py:87]: > Mh ... what if you have "blink_L" but you want to interpolate the data in left and right channels?

I would say "let's wait until someone wants to", or I would tell the person in this case to modify the list of channels annotated by the BAD_blink annotation. I think in practice if the person really did blink one eye, the other eye is open and doesn't need to be interpolated. So even if single-eye blinks are uncommon, we cover the "common" way this (uncommon) data should be processed I think

Comment (larsoner) [reply mne/preprocessing/eyetracking/eyetracking.py:87]: Probably -- worth trying. Also worth double checking that my understanding of the `ch_names` is correct...

But for this PR I think it doesn't matter too much. @britta-wstnr okay with adding / extending / figuring out how to deal with "I want to interpolate more channels than what the channel-aware annotations currently say are bad" later?

File: mne/preprocessing/eyetracking/eyetracking.py (hunk 2 above)
Comment (scott-huberty) [reply mne/preprocessing/eyetracking/eyetracking.py:31]: I agree that this reads better, thank you!

File: mne/preprocessing/eyetracking/eyetracking.py (hunk 3 above)
Comment (scott-huberty) [reply mne/preprocessing/eyetracking/eyetracking.py:14]: I think we included this in the "let's hold off for now and implement it if people ask for it". 

I think It's a little tough because for example if you only interpolate over some channels and not others then the data in the blink segments are still partly `"BAD"` (i.e. unusable samples or `np.nan` within the blink period).

I personally haven't had a scenario in my research where I've only wanted to interpolate some eyetrack channels and not others. What about you?

File: mne/preprocessing/eyetracking/eyetracking.py (hunk 4 above)
Comment (britta-wstnr) [reply mne/preprocessing/eyetracking/eyetracking.py:87]: > But for this PR I think it doesn't matter too much. @britta-wstnr okay with adding / extending / figuring out how to deal with "I want to interpolate more channels than what the channel-aware annotations currently say are bad" later?

Yes, yes, absolutely! This was just a question that formed in my mind while going over the code 🙂

File: mne/preprocessing/eyetracking/eyetracking.py (hunk 3 above)
Comment (britta-wstnr) [reply mne/preprocessing/eyetracking/eyetracking.py:14]: That's fair! I think I also need to remind myself that this is only for pupillometry channels to begin with (right?). Which actually makes me wonder if we should name the function more specifically to mark that - `interpolate_blinks_pupil` or something? We do not foresee this function to take over other channels as well, right?

Comment (larsoner) [reply mne/preprocessing/eyetracking/eyetracking.py:14]: I think the idea is that the `eyetracking` namespace it's nested under as `mne.preprocessing.eyetracking.interpolate_blinks` should be good enough to indicate the scope

Comment (scott-huberty) [reply mne/preprocessing/eyetracking/eyetracking.py:14]: @britta-wstnr You are totally right this is what we initially were going to do, but @larsoner and I met last week and felt that it's okay to interpolate over both `eyegaze` and `pupil` channels : )

Comment (britta-wstnr) [reply mne/preprocessing/eyetracking/eyetracking.py:14]: Mhhh I am not sure I agree interpolating `eyegaze` is a good idea - the position of the eye on the screen does not exist during a blink.

Comment (larsoner) [reply mne/preprocessing/eyetracking/eyetracking.py:14]: My argument was that the eye is still pointing somewhere in space and that's what the eye position/gaze is meant to reflect (even if it's blocked by an eyelid or something else). And it probably doesn't move during the blink / that is a reasonable assumption for most blinks.

Comment (larsoner) [reply mne/preprocessing/eyetracking/eyetracking.py:14]: ... I guess the question is what concrete use cases will be broken by interpolating? One that will be helped would be saying "the subject stayed fixed on a cross" or something so you can use `np.std` instead of `np.nanstd` etc. and it should still be valid. To me making people worry about `nan` when filtering, processing, etc. is not good for downstream stuff. My vote would be that the interpolation should interpolate everything by default at least, and we could add an option later to not do it if people end up needing it. But if that seems unsafe in some ways then we can do the reverse (though I'd like to add the `interpolate_gaze=False` *now* because I think I'd always want to set it to True in my analyses).

Comment (britta-wstnr) [reply mne/preprocessing/eyetracking/eyetracking.py:14]: Mh I don't think that is true - after all you can see blinks in EOG channels because the eye moves, right? -- A quick literature search indeed suggests that there is eye movement associated with blinking.
And especially if the gaze is expressed as position on the screen, an interpolation would be very misleading.

Comment (drammock) [reply mne/preprocessing/eyetracking/eyetracking.py:14]: > Mh I don't think that is true - after all you can see blinks in EOG channels because the eye moves, right? -- A quick literature search indeed suggests that there is eye movement associated with blinking.

I am with @britta-wstnr here, I thought that the eyes naturally roll upward during blinks

Comment (larsoner) [reply mne/preprocessing/eyetracking/eyetracking.py:14]: > Part of the decision as I remember was one of practicality.

So maybe we do add `interpolate_gaze=False` and if people want to do something where they know the gaze is technically undefined but want `epochs.average` and `.filter(picks='gaze')` or whatever to "just work" then they can do it.

Comment (britta-wstnr) [reply mne/preprocessing/eyetracking/eyetracking.py:14]: I wonder if the averaging would be something to fix in the averaging instead of interpolating (for gaze, that is).
Regarding the annotations: could the annotations be only removed from the pupil channels then? If I envision research paired with MEG and EEG, you'd want to retain information about blink periods anyway, even if you might want to interpolate your pupil data. I think it is not uncommon to clean MEEG data based on eye tracking information (e.g., I often kick out all trials that have a blink event in the eye tracking in my visual processing studies).

Comment (scott-huberty) [reply mne/preprocessing/eyetracking/eyetracking.py:14]: Alright sounds like the consensus is to not interpolate `eyegaze` channels by default -  I will add the `interpolate_gaze=False` flag and push a commit!


I agree that `np.nanmean` might be something we want to consider in the future for `epochs.average()` if this is an issue people keep running into!

**Comment by larsoner** [comment]
I think we should keep the annotation but adjust the interpolation function to just do nearest neighbor for those

**Comment by scott-huberty** [comment]
Ok just pushed a commit, a couple notes:

I just noticed that  `scipy.interpolation.interp1d`, which I was using, [appears to be considered legacy API](https://docs.scipy.org/doc/scipy/tutorial/interpolate/1D.html#legacy-interface-for-1-d-interpolation-interp1d) and is no longer recommended. 

For a simple linear interpolation, `numpy.interp`  seems to be the recommendation, and it has the added benefit that if you need to interpolate a value that is at the start or end of your array, `numpy.interp` will just use a zero order hold interpolation on that index (i.e. it handles the aforementioned issue where a blink was at the end of the recording).

**Review by larsoner — COMMENTED** [review]
Just some minor stuff from me, looks good!

File: tutorials/preprocessing/90_eyetracking_data.py [hunk 5]
```diff
@@ -142,25 +143,82 @@
...
+# %%
+# Dealing with artifacts
+# ----------------------
+# From the plot above, we see that there are some artifacts in the data that we should
+# remove before analyzing the pupil response. First, we notice that there is some
+# high frequency noise in the pupil signal, likely due to the sub-optimal calibration
+# of the eye tracker. We can remove this noise by low-pass filtering the data:
+
+# Apply a low pass filter to the pupil channel
+raw.filter(l_freq=None, h_freq=40, picks=["pupil_right"])
```
[the remark this answers is not in the document]
Comment (scott-huberty) [reply tutorials/preprocessing/90_eyetracking_data.py:155]: values for pupil signal are `0` during blinks (at least, for EyeLink data..), but I get your point and so will re-order!

File: tutorials/preprocessing/90_eyetracking_data.py [hunk 6]
```diff
@@ -115,7 +114,9 @@
 # the screen. We now extract these events to visualize the pupil response. We will use
 # these later in this tutorial.
 
-events = find_events(raw, "DIN", shortest_event=1, min_duration=0.02, uint_cast=True)
+events = mne.find_events(
+    raw, "DIN", shortest_event=1, min_duration=0.02, uint_cast=True
```
[the remark this answers is not in the document]
Comment (scott-huberty) [reply tutorials/preprocessing/90_eyetracking_data.py:118]: Yeah it is the only STIM_CH int this object, I just checked and can confirm that you are right, we don't need to explicitly specify "DIN" here

File: tutorials/preprocessing/90_eyetracking_data.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (scott-huberty) [reply tutorials/preprocessing/90_eyetracking_data.py:155]: Just realized that your point doesn't only apply to  `np.nan`. Again will do this after interpolation.

**Comment by larsoner** [comment]
Thanks @scott-huberty !
